core/src/main/java/org/apache/calcite/rel/rules/PruneEmptyRules.java

```diff
@@ -480,9 +486,25 @@ public interface JoinLeftEmptyRuleConfig extends PruneEmptyRule.Config {
       return new PruneEmptyRule(this) {
         @Override public void onMatch(RelOptRuleCall call) {
           Join join = call.rel(0);
+          Values empty = call.rel(1);
+          RelNode right = call.rel(2);
+          RexBuilder rexBuilder = call.builder().getRexBuilder();
           if (join.getJoinType().generatesNullsOnLeft()) {
             // "select * from emp right join dept" is not necessarily empty if
             // emp is empty
+            // join can be removed and take the right branch, all columns come from emp are null
+            List<RexNode> projects = new ArrayList<>(
+                empty.getRowType().getFieldCount() + right.getRowType().getFieldCount());
+            List<String> columnNames = new ArrayList<>(
+                empty.getRowType().getFieldCount() + right.getRowType().getFieldCount());
+            // left
+            addNullLiterals(rexBuilder, empty, projects, columnNames);
+            // right
+            copyProjects(rexBuilder, right.getRowType(),
+                join.getRowType(), empty.getRowType().getFieldCount(), projects, columnNames);
+
+            RelNode project = call.builder().push(right).project(projects, columnNames).build();
+            call.transformTo(project);
```

Comment:
Can this be written in a functional style? Rather than creating empty lists then filling them, use `map`. `RelBuilder` has plenty of methods that return lists to types, lists of names, etc.

Reply:
Refactored the utility methods to use the Java stream api. I defined stream of List<Pair<>> as a return type to ensure the order of expressions.
If this is not what you mean could you please give some example?